Add a gzip indexing class for faster gz navigation

phord opened the pull request:
The gzread function is slow.  Every time you seek to a new location, the
whole file up to that position has to be decompressed again.  This causes
massive lags when trying to do simple things in lnav on a large .gz file.

Use the zlib inflate* functions instead and record the dictionary
periodically while processing the file the first time.  Then use
inflateSetDictionary to restore the dictionary to a convenient
location when trying to seek into the file again in the future.

Use a default period of 1MB of compressed data for syncpoints.
Each syncpoint uses 32KB. This is a ratio of 3.2%.  For example,
a 1GB .gz file (compressed size) will require us to keep 32MB
of index data in memory. A better method may be to use a fixed
number of syncpoints and divide the file appropriately. This
would keep the memory bounded at the cost of slower file
navigation on large .gz files.

Use pread to read the data for the stream decompressor and remove
the lock_hack previously employed.

NB. The documentation on these zlib functions is sparse. I followed
the example in zlib/examples/zran.c, but I used the z_stream total_in
and total_out variables instead of keeping my own separately as zran.c
does.  Maybe this is incompatible with some very old zlib versions.
I haven't looked.

tstack commented:
Neato!  Not sure why travis bombed.  I think the approach is sound, I'll take a closer look at the changes a bit later.

phord commented:
Fixed the travis breakage by giving the new class a move-ctor.  Without it, it broke the move-ctor for line_buffer.

tstack commented on src/line_buffer.cc:
```diff
@@ -116,9 +116,154 @@ class lock_hack {
+        inflateEnd(&this->strm);
+        ::close(this->gz_fd);
+        this->syncpoints.clear();
+        this->gz_fd = 0;
```
Can you use -1 for an uninitialized fd (or nonstd::optional<int>)

tstack commented on src/line_buffer.cc:
```diff
@@ -116,9 +116,154 @@ class lock_hack {
+
+void line_buffer::gz_indexed::init_stream()
+{
+    if (*this) inflateEnd(&this->strm);
```
nit: add curly brackets

tstack commented on src/line_buffer.cc:
```diff
@@ -116,9 +116,154 @@ class lock_hack {
+    this->strm.avail_out = size;
+    this->strm.next_out = (unsigned char *) buf;
+
+    int last = this->syncpoints.size() ?
```
nit: use empty() instead of size() (?)

tstack commented on src/line_buffer.cc:
```diff
@@ -116,9 +116,154 @@ class lock_hack {
+
+            if (this->strm.total_in >= last + SYNCPOINT_SIZE &&
+                size > this->strm.avail_out + GZ_WINSIZE &&
+                (strm.data_type & 128) && !(strm.data_type & 64)) {
```
nit: can you add some constants for the 128/64 magic numbers

phord replied:
I tried to use `gzip.h` constants for these, but they seem to be nonexistent.  Disappointing.  So I used the code example from libz.  But I agree, these suck.  Will fix.

tstack commented on src/line_buffer.hh:
```diff
@@ -74,6 +75,85 @@ public:
+                assert((s.data_type & 128));
+                assert(!(s.data_type & 64));
+                assert(size >= s.avail_out + GZ_WINSIZE);
+                this->bits = s.data_type & 7;
```
nit: 7 magic number

tstack commented:
Can you also add some test cases to test_line_buffer2.cc/test_line_buffer.sh

tstack commented:
@phord Are you going to have time to make these updates?

phord commented:
Yep. Sorry I didn't see these earlier.

On Fri, Aug 9, 2019, 9:18 AM Tim Stack <notifications@github.com> wrote:

> @phord <https://github.com/phord> Are you going to have time to make
> these updates?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tstack/lnav/pull/677?email_source=notifications&email_token=AAA6IBHXARMEZGM4VWDACSDQDWKE5A5CNFSM4IIXJAKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD37D7SI#issuecomment-519978953>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA6IBDPVEXDGB2BTXVFHKTQDWKE5ANCNFSM4IIXJAKA>
> .
>

phord commented:
Just starting to look at the test code.  I'll see about adding tests later this week.

phord commented:
Added tests to verify random access and concatenated .gz files.  The 2nd one didn't actually work, so I added a test for it.

For the random access test, I used drive_line_buffer.  It doesn't support comparing to a .gz file as source, so I added a 2nd argument to pass in a decompressed file to compare.

tstack commented:
thanks!
